go/ql/src/experimental/CWE-347/JWTParsingAlgorithm.ql

```diff
@@ -0,0 +1,57 @@
+/**
+ * @name JWT Method Check
+ * @description Trusting the Method provided by the incoming JWT token may lead to an algorithim confusion
+ * @kind problem
+ * @problem.severity error
+ * @security-severity 8.0
+ * @precision medium
+ * @id go/jwt-alg-confusion
+ * @tags security
+ *       external/cwe/cwe-347
+ */
+
+import go
+import experimental.frameworks.JWT
+import DataFlow
+
+/**
+ * A parse function that verifies signature and accepts all methods.
+ */
+class SafeJwtParserFunc extends Function {
+  SafeJwtParserFunc() { this.hasQualifiedName(golangJwtModern(), ["Parse", "ParseWithClaims"]) }
+}
+
+/**
+ * A parse method that verifies signature.
+ */
+class SafeJwtParserMethod extends Method {
+  SafeJwtParserMethod() {
+    this.hasQualifiedName(golangJwtModern(), "Parser", ["Parse", "ParseWithClaims"])
+  }
+}
+
+from CallNode c, Function func
+where
+  (
+    c.getTarget() = func and
+    // //Flow from NewParser to Parse (check that this call to Parse does not use a Parser that sets Valid Methods)
+    (
+      func instanceof SafeJwtParserMethod and
+      not exists(CallNode c2, WithValidMethods wvm, NewParser m |
+        c2.getTarget() = m and
+        (
+          c2.getCall().getAnArgument() = wvm.getACall().asExpr() and
+          DataFlow::localFlow(c2.getAResult(), c.getReceiver())
+        )
+      )
+      or
+      //ParserFunc creates a new default Parser on call that accepts all methods
+      func instanceof SafeJwtParserFunc
+    ) and
+    //Check that the Parse(function or method) does not check the Token Method field, which most likely is a check for method type
+    not exists(Field f |
+      f.hasQualifiedName(golangJwtModern(), "Token", "Method") and
+      f.getARead().getRoot() = c.getCall().getAnArgument()
```

Comment:
Is it possible to specify which argument?

Reply:
Also, I think this only works if the function literal is specified in place. If it's assigned to a variable and then that variable is used as an argument, would this catch it?

Reply:
@owen-mc having a bit of trouble coming up with a good way to see if the function name that is passed into the argument has a field read of the Method or not. I can get rid of the `Var only used in one side of disjunct` error but if you can find another way that avoids strings altogether please let me know.